src/jit/hwintrinsicxarch.cpp

```diff
@@ -152,6 +158,152 @@ InstructionSet Compiler::isaOfHWIntrinsic(NamedIntrinsic intrinsic)
+//         - from the hwIntrinsicInfoArray table if intrinsic is NOT HW_Category_UnfixedSIMDSize
+//         - TODO - from the signature if intrinsic is HW_Category_UnfixedSIMDSize
+//
+// Note - this function is only used by the importer
+//        after importation (i.e., codegen), we can get the SIMD size from GenTreeHWIntrinsic IR
+static unsigned simdSizeOfHWIntrinsic(NamedIntrinsic intrinsic, CORINFO_SIG_INFO* sig)
+{
+    assert(intrinsic > NI_HW_INTRINSIC_START && intrinsic < NI_HW_INTRINSIC_END);
+    return hwIntrinsicInfoArray[intrinsic - NI_HW_INTRINSIC_START - 1].simdSize;
+}
+
+//------------------------------------------------------------------------
+// numArgsOfHWIntrinsic: get the number of arguments
+//
+// Arguments:
+//    intrinsic -- id of the intrinsic function.
+//
+// Return Value:
+//     number of arguments
+//
+int Compiler::numArgsOfHWIntrinsic(NamedIntrinsic intrinsic)
+{
+    assert(intrinsic != NI_Illegal);
+    assert(intrinsic > NI_HW_INTRINSIC_START && intrinsic < NI_HW_INTRINSIC_END);
+    return hwIntrinsicInfoArray[intrinsic - NI_HW_INTRINSIC_START - 1].numArgs;
+}
+
+//------------------------------------------------------------------------
+// insOfHWIntrinsic: get the instruction of the given intrinsic
+//
+// Arguments:
+//    intrinsic -- id of the intrinsic function.
+//    type      -- vector base type of this intrinsic
+//
+// Return Value:
+//     the instruction of the given intrinsic on the base type
+//     return INS_invalid for unsupported base types
+//
+instruction Compiler::insOfHWIntrinsic(NamedIntrinsic intrinsic, var_types type)
+{
+    assert(intrinsic != NI_Illegal);
+    assert(intrinsic > NI_HW_INTRINSIC_START && intrinsic < NI_HW_INTRINSIC_END);
+    assert(type >= TYP_BYTE && type <= TYP_DOUBLE);
+    return hwIntrinsicInfoArray[intrinsic - NI_HW_INTRINSIC_START - 1].ins[type - TYP_BYTE];
+}
+
+//------------------------------------------------------------------------
+// categoryOfHWIntrinsic: get the category of the given intrinsic
+//
+// Arguments:
+//    intrinsic -- id of the intrinsic function.
+//    type      -- vector base type of this intrinsic
+//
+// Return Value:
+//     the category of the given intrinsic
+//
+HWIntrinsicCategory Compiler::categoryOfHWIntrinsic(NamedIntrinsic intrinsic)
+{
+    assert(intrinsic != NI_Illegal);
+    assert(intrinsic > NI_HW_INTRINSIC_START && intrinsic < NI_HW_INTRINSIC_END);
+    return hwIntrinsicInfoArray[intrinsic - NI_HW_INTRINSIC_START - 1].category;
+}
+
+//------------------------------------------------------------------------
+// getArgForHWIntrinsic: get the argument from the stack and match  the signature
+//
+// Arguments:
+//    sig   -- signature
+//    index -- the index of the argument in the signature
+//
+// Return Value:
+//     get the argument at the given index from the stack and match  the signature
+//
+GenTree* Compiler::getArgForHWIntrinsic(CORINFO_SIG_INFO* sig, unsigned index)
```

Comment:
I'm not sure I like this method. It seems very generalized (possibly over-generalized) and doesn't give a clear picture of what we expect certain types/args to be.

It was much easier to read/follow and was simple to know what to expect for a given intrinsic when we had it in the form of;
```
switch (intrinsicID)
{
    case NI_*:  // These take two SIMD arguments and return a SIMD argument
        op2 = impSIMDPopStack();
        op1 = impSIMDPopStack();
        retNode = gtNewHWIntrinsicNode();
}
```

Using categories would allow the same, but without so many entries in the switch table.

Reply:
IMO @tannergooding proposal is easier to follow when writing and reading code

Reply:
I am trying to handle Vector/non-vector arguments in the table-driven framework.

Reply:
> I am trying to use this function to handle Vector/non-vector arguments in the table-driven framework.

Right. I was just indicating that I find it significantly harder to read in this form and it might be better to have the slightly more verbose option for increased maintainability (This is just my opinion though, others may prefer this better).